metricbeat/module/vsphere/virtualmachine/virtualmachine.go

```diff
@@ -104,115 +130,277 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+	vmList, err := m.getVirtualMachineList(ctx, c, mgr, rpList)
 	if err != nil {
-		return fmt.Errorf("error in Retrieve: %w", err)
+		return fmt.Errorf("error getting Virtual Machine List: %w", err)
 	}
 
-	for _, vm := range vmt {
-		usedMemory := int64(vm.Summary.QuickStats.GuestMemoryUsage) * 1024 * 1024
-		usedCPU := vm.Summary.QuickStats.OverallCpuUsage
-		event := mapstr.M{
-			"name": vm.Summary.Config.Name,
-			"os":   vm.Summary.Config.GuestFullName,
-			"cpu": mapstr.M{
-				"used": mapstr.M{
-					"mhz": usedCPU,
-				},
-			},
-			"memory": mapstr.M{
-				"used": mapstr.M{
-					"guest": mapstr.M{
-						"bytes": usedMemory,
-					},
-					"host": mapstr.M{
-						"bytes": int64(vm.Summary.QuickStats.HostMemoryUsage) * 1024 * 1024,
-					},
-				},
-			},
+	vmChunks := m.getVmChunks(vmList)
+	vmMap := getVmMap(vmList)
+
+	for _, chunk := range vmChunks {
+		// Perform the query for all VMs at once
+		metricBase, err := perfManager.SampleByName(ctx, querySpec, metricsToQuery, chunk)
+		if err != nil {
+			return fmt.Errorf("Error retrieving metrics: %s\n", err)
 		}
 
-		totalCPU := vm.Summary.Config.CpuReservation
-		if totalCPU > 0 {
-			freeCPU := totalCPU - usedCPU
-			// Avoid negative values if reported used CPU is slightly over total configured.
-			if freeCPU < 0 {
-				freeCPU = 0
+		// Assign metrics to the event output
+		for _, base := range metricBase {
+			metric, ok := base.(*types.PerfEntityMetric)
+			if !ok {
+				continue
 			}
-			event.Put("cpu.total.mhz", totalCPU)
-			event.Put("cpu.free.mhz", freeCPU)
-		}
 
-		totalMemory := int64(vm.Summary.Config.MemorySizeMB) * 1024 * 1024
-		if totalMemory > 0 {
-			freeMemory := totalMemory - usedMemory
-			// Avoid negative values if reported used memory is slightly over total configured.
-			if freeMemory < 0 {
-				freeMemory = 0
+			// Build initial event object with static values
+			virtualMachine := vmMap[metric.Entity.Value]
+			event := mapstr.M{
+				"name":             virtualMachine.Summary.Config.Name,
+				"os":               virtualMachine.Summary.Config.GuestFullName,
+				"os_family":        virtualMachine.Guest.GuestFamily,
+				"guest_state":      virtualMachine.Guest.GuestState,
+				"hardware_version": virtualMachine.Guest.HwVersion,
+				"vmtools_version":  virtualMachine.Guest.ToolsVersion,
+				"uuid":             virtualMachine.Summary.Config.InstanceUuid,
+				"id":               virtualMachine.Reference().Value,
+				"primary_ip":       virtualMachine.Guest.IpAddress,
+				"resource_pool": mapstr.M{
+					"id": virtualMachine.ResourcePool.Value,
+				},
+				"cpu": mapstr.M{
+					"reserved": virtualMachine.Summary.Config.CpuReservation,
+					"cores":    virtualMachine.Summary.Config.NumCpu,
+				},
+				"disks": mapstr.M{
+					"count": virtualMachine.Summary.Config.NumVirtualDisks,
+				},
+				"storage": mapstr.M{
+					"committed":  virtualMachine.Summary.Storage.Committed,
+					"uncommited": virtualMachine.Summary.Storage.Uncommitted,
```

Comment:
correct the spelling of uncommited
```suggestion
					"uncommitted": virtualMachine.Summary.Storage.Uncommitted,
```

Reply:
fixed